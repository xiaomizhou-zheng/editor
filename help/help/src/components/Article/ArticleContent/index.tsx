import React from 'react';
import UIAnalyticsEvent from '@atlaskit/analytics-next/UIAnalyticsEvent';

import { HistoryItem, ARTICLE_TYPE } from '../../../model/Help';
import { REQUEST_STATE } from '../../../model/Requests';
import { Article as ArticleType } from '../../../model/Article';
import { WhatsNewArticle as WhatsNewArticleType } from '../../../model/WhatsNew';

import HelpArticle from '../HelpArticle';
import ArticleLoadingFail from '../ArticleLoadingFail';
import WhatsNewArticle from '../WhatsNewArticle';

interface Props {
  currentArticle: HistoryItem | undefined;
  onHelpArticleLoadingFailTryAgainButtonClick?(
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    analyticsEvent: UIAnalyticsEvent,
  ): void;
  onWhatsNewArticleLoadingFailTryAgainButtonClick?(
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    analyticsEvent: UIAnalyticsEvent,
  ): void;
}

export const ArticleContent: React.FC<Props> = ({
  currentArticle,
  onHelpArticleLoadingFailTryAgainButtonClick,
  onWhatsNewArticleLoadingFailTryAgainButtonClick,
}) => {
  if (currentArticle && currentArticle.id) {
    if (currentArticle.type === ARTICLE_TYPE.HELP_ARTICLE) {
      const article = currentArticle.article as ArticleType;
      if (article && currentArticle.state === REQUEST_STATE.done) {
        return <HelpArticle article={article} />;
      } else if (currentArticle.state === REQUEST_STATE.loading) {
        return <HelpArticle isLoading />;
      } else if (currentArticle.state === REQUEST_STATE.error) {
        return (
          <ArticleLoadingFail
            onTryAgainButtonClick={onHelpArticleLoadingFailTryAgainButtonClick}
          />
        );
      }
    }

    if (currentArticle.type === ARTICLE_TYPE.WHATS_NEW) {
      const article = currentArticle.article as WhatsNewArticleType;
      if (article && currentArticle.state === REQUEST_STATE.done) {
        return <WhatsNewArticle article={article} />;
      } else if (currentArticle.state === REQUEST_STATE.loading) {
        return <WhatsNewArticle isLoading />;
      } else if (currentArticle.state === REQUEST_STATE.error) {
        return (
          <ArticleLoadingFail
            onTryAgainButtonClick={
              onWhatsNewArticleLoadingFailTryAgainButtonClick
            }
          />
        );
      }
    }
  }

  return null;
};

export default ArticleContent;
