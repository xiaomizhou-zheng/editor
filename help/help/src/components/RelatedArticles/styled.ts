/** @jsx jsx */

import { css } from '@emotion/core';
import styled from '@emotion/styled';
import * as colors from '@atlaskit/theme/colors';
import { gridSize, fontSize } from '@atlaskit/theme/constants';
import { token } from '@atlaskit/tokens';

const baseHeading = (size: number, lineHeight: number) => `
  font-size: ${size / fontSize()}em;
  font-style: inherit;
  line-height: ${lineHeight / size};
`;

export const truncate = (width: string = '100%') => css`
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: ${width};
`;

export const RelatedArticlesTitle = styled.div`
  ${baseHeading(16, 20)}
  color: ${token('color.text', colors.N800)};
  font-weight: 600;
  letter-spacing: -0.006em;
  padding: ${gridSize() * 2}px 0;
`;

/**
 * Loading styled-components
 */
export const LoadignRelatedArticleSection = styled.div`
  margin-top: ${gridSize()}px;
`;

export const LoadignRelatedArticleList = styled.ul`
  width: 100%;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
`;

export const LoadignRelatedArticleListItem = styled.li`
  display: block;
  width: 100%;
  padding: ${gridSize()}px;
  margin-bottom: ${gridSize() * 2}px;
  box-sizing: border-box;
`;
