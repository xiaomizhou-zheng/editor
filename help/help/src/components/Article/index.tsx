import React, {
  useState,
  useLayoutEffect,
  useRef,
  useEffect,
  useCallback,
} from 'react';
import { Transition } from 'react-transition-group';
import UIAnalyticsEvent from '@atlaskit/analytics-next/UIAnalyticsEvent';

import { useNavigationContext } from '../contexts/navigationContext';
import { useHelpArticleContext } from '../contexts/helpArticleContext';
import {
  SLIDEIN_OVERLAY_TRANSITION_DURATION_MS,
  TRANSITION_STATUS,
  VIEW,
} from '../constants';
import { ArticleContainer } from './styled';
import ArticleContent from './ArticleContent';
import type { HistoryItem } from '../../model/Help';

// Animation
const defaultStyle = {
  left: `100%`,
};

const enableTransition: { [id: string]: React.CSSProperties } = {
  enabled: {
    transition: `left ${SLIDEIN_OVERLAY_TRANSITION_DURATION_MS}ms cubic-bezier(0.2, 0, 0, 1) 0s`,
  },
};

const transitionStyles: { [id: string]: React.CSSProperties } = {
  entered: { left: 0 },
  exited: { left: `100%` },
};

export const Article: React.FC = () => {
  const {
    view,
    articleId,
    history,
    getCurrentArticle,
    reloadHelpArticle,
    reloadWhatsNewArticle,
  } = useNavigationContext();
  const {
    onHelpArticleLoadingFailTryAgainButtonClick,
  } = useHelpArticleContext();

  const [
    skipArticleSlideInAnimation,
    setSkipArticleSlideInAnimation,
  ] = useState<boolean>(
    articleId?.id !== '' || (history ? history.length : []) > 0,
  );
  const [showArticle, setShowArticle] = useState<boolean>(
    skipArticleSlideInAnimation,
  );

  const currentArticleValue: HistoryItem | undefined = getCurrentArticle();
  const [currentArticle, setCurrentArticle] = useState<HistoryItem | undefined>(
    currentArticleValue,
  );

  const articleContainerRef = useRef<HTMLDivElement>(null);
  const onArticleEnteredTimeout = useRef<number>();
  const onCurrentArticleChangedTimeout = useRef<number>();

  useLayoutEffect(() => {
    if (articleContainerRef.current) {
      articleContainerRef.current.scrollTop = 0;
    }
  }, [history]);

  const onArticleEntered = (): void => {
    onArticleEnteredTimeout.current = window.setTimeout(() => {
      // if skipArticleSlideInAnimation is true, set to false after the
      // first slide-in animation
      // NOTE: skipArticleSlideInAnimation could be true only after the mounting
      if (skipArticleSlideInAnimation) {
        setSkipArticleSlideInAnimation(false);
      }
    }, SLIDEIN_OVERLAY_TRANSITION_DURATION_MS);
  };

  const onArticleExit = (): void => {
    clearTimeout(onArticleEnteredTimeout.current);
  };

  const handleOnHelpArticleLoadingFailTryAgainButtonClick = useCallback(
    (
      event: React.MouseEvent<HTMLElement, MouseEvent>,
      analyticsEvent: UIAnalyticsEvent,
    ): void => {
      if (onHelpArticleLoadingFailTryAgainButtonClick) {
        onHelpArticleLoadingFailTryAgainButtonClick(
          event,
          analyticsEvent,
          articleId,
        );
      }

      if (reloadHelpArticle && currentArticle) {
        reloadHelpArticle(currentArticle);
      }
    },
    [
      articleId,
      currentArticle,
      onHelpArticleLoadingFailTryAgainButtonClick,
      reloadHelpArticle,
    ],
  );

  const handleOnWhatsNewArticleLoadingFailTryAgainButtonClick = useCallback(
    (
      event: React.MouseEvent<HTMLElement, MouseEvent>,
      analyticsEvent: UIAnalyticsEvent,
    ): void => {
      if (onHelpArticleLoadingFailTryAgainButtonClick) {
        onHelpArticleLoadingFailTryAgainButtonClick(
          event,
          analyticsEvent,
          articleId,
        );
      }
      if (reloadWhatsNewArticle && currentArticle) {
        reloadWhatsNewArticle(currentArticle);
      }
    },
    [
      currentArticle,
      articleId,
      onHelpArticleLoadingFailTryAgainButtonClick,
      reloadWhatsNewArticle,
    ],
  );

  useEffect(() => {
    clearTimeout(onCurrentArticleChangedTimeout.current);
    if (currentArticleValue) {
      setCurrentArticle(currentArticleValue);
    } else {
      onCurrentArticleChangedTimeout.current = window.setTimeout(() => {
        setCurrentArticle(currentArticleValue);
      }, SLIDEIN_OVERLAY_TRANSITION_DURATION_MS);
    }
  }, [currentArticleValue]);

  useEffect(() => {
    if (showArticle) {
      setShowArticle(
        view === VIEW.ARTICLE ||
          view === VIEW.SEARCH ||
          view === VIEW.WHATS_NEW_ARTICLE,
      );
    } else {
      setShowArticle(view === VIEW.ARTICLE || view === VIEW.WHATS_NEW_ARTICLE);
    }
  }, [view, showArticle]);

  return (
    <Transition
      in={showArticle}
      timeout={SLIDEIN_OVERLAY_TRANSITION_DURATION_MS}
      enter={!skipArticleSlideInAnimation}
      onEntered={onArticleEntered}
      onExit={onArticleExit}
      unmountOnExit
      mountOnEnter
    >
      {(state: TRANSITION_STATUS) => {
        return (
          <ArticleContainer
            ref={articleContainerRef}
            style={{
              ...defaultStyle,
              ...transitionStyles[state],
              ...enableTransition[
                !skipArticleSlideInAnimation ? 'enabled' : 'disabled'
              ],
            }}
          >
            <ArticleContent
              currentArticle={currentArticle}
              onHelpArticleLoadingFailTryAgainButtonClick={
                reloadHelpArticle &&
                handleOnHelpArticleLoadingFailTryAgainButtonClick
              }
              onWhatsNewArticleLoadingFailTryAgainButtonClick={
                reloadWhatsNewArticle &&
                handleOnWhatsNewArticleLoadingFailTryAgainButtonClick
              }
            />
          </ArticleContainer>
        );
      }}
    </Transition>
  );
};

export default Article;
