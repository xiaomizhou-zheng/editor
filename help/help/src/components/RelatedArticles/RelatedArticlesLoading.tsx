import React from 'react';
import { injectIntl, WrappedComponentProps } from 'react-intl-next';

import { messages } from '../../messages';

import { DividerLine, LoadingRectangle } from '../../util/styled';

import {
  LoadignRelatedArticleList,
  LoadignRelatedArticleListItem,
} from './styled';

const RelatedArticlesLoading: React.FC<WrappedComponentProps> = ({
  intl: { formatMessage },
}) => {
  return (
    <>
      <LoadignRelatedArticleList
        aria-label={formatMessage(messages.help_loading)}
        role="img"
      >
        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>

        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>

        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>
        <DividerLine />
      </LoadignRelatedArticleList>
    </>
  );
};

export default injectIntl(RelatedArticlesLoading);
