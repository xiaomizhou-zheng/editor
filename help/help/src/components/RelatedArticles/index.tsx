import React, { useState, useEffect, useCallback } from 'react';
import {
  UIAnalyticsEvent,
  withAnalyticsContext,
} from '@atlaskit/analytics-next';
import SectionMessage from '@atlaskit/section-message';
import Button from '@atlaskit/button/custom-theme-button';
import { injectIntl, WrappedComponentProps } from 'react-intl-next';

import {
  name as packageName,
  version as packageVersion,
} from '../../version.json';
import { messages } from '../../messages';
import { ArticleItem } from '../../model/Article';

import ArticlesList from '../ArticlesList';

import RelatedArticlesLoading from './RelatedArticlesLoading';
import { DividerLine } from '../../util/styled';
import { RelatedArticlesTitle } from './styled';
import useCancellablePromise from '../../util/hooks/cancellablePromise';
import { usePrevious } from '../../util/hooks/previous';

export interface Props {
  // Style. This component has two different styles (primary and secondary)
  style?: 'primary' | 'secondary';
  // routeGroup used to get the related articles. This prop is optional.
  routeGroup?: string;
  // routeName used to get the related articles. This prop is optional.
  routeName?: string;
  // Function used to get related articles. This prop is optional, if is not defined the related articles will not be displayed
  onGetRelatedArticles?(
    routeGroup?: string,
    routeName?: string,
  ): Promise<ArticleItem[]>;
  /* Function executed when the user clicks on of the related articles */
  onRelatedArticlesListItemClick?: (
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    analyticsEvent: UIAnalyticsEvent,
    articleData: ArticleItem,
  ) => void;
  /* Function executed when the user clicks "Show More" */
  onRelatedArticlesShowMoreClick?: (
    event: React.MouseEvent<HTMLElement>,
    analyticsEvent: UIAnalyticsEvent,
    isCollapsed: boolean,
  ) => void;
}

export const RelatedArticles: React.FC<Props & WrappedComponentProps> = ({
  style = 'primary',
  routeGroup,
  routeName,
  onGetRelatedArticles,
  onRelatedArticlesListItemClick,
  onRelatedArticlesShowMoreClick,
  intl: { formatMessage },
}) => {
  const { cancellablePromise } = useCancellablePromise();
  const [relatedArticles, setRelatedArticles] = useState<ArticleItem[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [hasError, setHasError] = useState<boolean>(false);
  const prevRouteGroup = usePrevious(routeGroup);
  const prevRouteName = usePrevious(routeName);

  const handleOnRelatedArticlesShowMoreClick = (
    event: React.MouseEvent<HTMLElement>,
    analyticsEvent: UIAnalyticsEvent,
    isCollapsed: boolean,
  ) => {
    analyticsEvent.payload.attributes = {
      componentName: 'RelatedArticles',
      packageName,
      packageVersion,
    };
    if (onRelatedArticlesShowMoreClick) {
      onRelatedArticlesShowMoreClick(event, analyticsEvent, isCollapsed);
    }
  };

  const updateRelatedArticles = useCallback(async () => {
    if (onGetRelatedArticles) {
      try {
        const relatedArticles: ArticleItem[] = await cancellablePromise(
          onGetRelatedArticles(routeGroup, routeName),
        );
        setRelatedArticles(relatedArticles);
        setIsLoading(false);
        setHasError(false);
      } catch (error) {
        setHasError(true);
      }
    } else {
      setIsLoading(false);
      setHasError(false);
    }
  }, [cancellablePromise, onGetRelatedArticles, routeGroup, routeName]);

  useEffect(() => {
    if (routeGroup !== prevRouteGroup || routeName !== prevRouteName) {
      updateRelatedArticles();
    }
  }, [
    prevRouteGroup,
    prevRouteName,
    routeGroup,
    routeName,
    updateRelatedArticles,
  ]);

  if (hasError) {
    return (
      <SectionMessage appearance="warning">
        <p>
          <strong>
            {formatMessage(messages.help_related_article_endpoint_error_title)}
          </strong>
        </p>
        <p>
          {formatMessage(
            messages.help_related_article_endpoint_error_description,
          )}
        </p>
        <p>
          <Button
            appearance="link"
            spacing="compact"
            css={{ padding: '0', '& span': { margin: '0' } }}
            onClick={updateRelatedArticles}
          >
            {formatMessage(
              messages.help_related_article_endpoint_error_button_label,
            )}
          </Button>
        </p>
      </SectionMessage>
    );
  } else {
    return (
      <>
        {style === 'secondary' && relatedArticles.length > 0 && (
          <>
            <DividerLine />
            <RelatedArticlesTitle>
              {formatMessage(messages.help_related_article_title)}
            </RelatedArticlesTitle>
          </>
        )}
        {isLoading ? (
          <RelatedArticlesLoading />
        ) : (
          <ArticlesList
            style={style}
            articles={relatedArticles}
            onArticlesListItemClick={onRelatedArticlesListItemClick}
            onToggleArticlesList={handleOnRelatedArticlesShowMoreClick}
          />
        )}
      </>
    );
  }
};

export default withAnalyticsContext({
  componentName: 'RelatedArticles',
  packageName,
  packageVersion,
})(injectIntl(RelatedArticles));
