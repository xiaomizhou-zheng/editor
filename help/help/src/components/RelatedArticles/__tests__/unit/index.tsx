import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render, waitFor, fireEvent } from '@testing-library/react';
import { createIntl, createIntlCache, IntlProvider } from 'react-intl-next';

import {
  getMockArticleItemList,
  getMockArticleItem,
} from '../../../../util/testing/mock';
import { messages } from '../../../../messages';

import RelatedArticles from '../../index';
import { ArticleItem } from '../../../../model/Article';

// Messages
const cache = createIntlCache();
const intl = createIntl(
  {
    locale: 'en',
    messages: {},
  },
  cache,
);
const messageEndpointErrorTitle = intl.formatMessage(
  messages.help_related_article_endpoint_error_title,
);
const messageEndpointErrorDescription = intl.formatMessage(
  messages.help_related_article_endpoint_error_description,
);
const messageEndpointErrorButtonLabel = intl.formatMessage(
  messages.help_related_article_endpoint_error_button_label,
);
const messageTitle = intl.formatMessage(messages.help_related_article_title);
const messageShowMore = intl.formatMessage(
  messages.help_show_more_button_label_more,
  { numberOfItemsLeft: '2', itemsType: 'articles' },
);
const messageLoading = intl.formatMessage(messages.help_loading);

const NUMBER_OF_ARTICLES = 5;
const mockArticleItem = getMockArticleItem('1');
const relativeArticlesPromise: (
  numberOfItems: number,
) => Promise<ArticleItem[]> = (numberOfItems = 0) =>
  new Promise((resolve, reject) => {
    if (numberOfItems > 0) {
      setTimeout(() => {
        resolve(getMockArticleItemList(numberOfItems));
      }, 100);
    } else {
      reject('error');
    }
  });

const mockOnGetRelatedArticles = jest
  .fn()
  .mockImplementation(relativeArticlesPromise);
const mockOnRelatedArticlesListItemClick = jest.fn();
const mockOnRelatedArticlesShowMoreClick = jest.fn();

describe('RelatedArticles', () => {
  jest.useFakeTimers();

  beforeEach(() => {
    jest.clearAllTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should match snapshot', async () => {
    const { container } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should request the related article on the first render', async () => {
    render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(mockOnGetRelatedArticles).toHaveBeenCalledTimes(1);
  });

  it('Should show loading state', async () => {
    const { getByLabelText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(getByLabelText(messageLoading)).not.toBeNull();
  });

  it('Should display the related articles', async () => {
    const { getAllByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(getAllByText(mockArticleItem.title).length).toBeGreaterThan(0);
  });

  it('Should execute "onRelatedArticlesListItemClick" when the user clicks one of the list items', async () => {
    const { getAllByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    const firstItem = getAllByText(mockArticleItem.title)[0].closest('a');

    expect(firstItem).not.toBeNull();

    if (firstItem) {
      expect(mockOnRelatedArticlesListItemClick).toHaveBeenCalledTimes(0);
      fireEvent.click(firstItem);
      expect(mockOnRelatedArticlesListItemClick).toHaveBeenCalledTimes(1);
    }
  });

  it('Should execute "mockOnRelatedArticlesShowMoreClick" when the user clicks the "Show more" button', async () => {
    const { getByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    const showMoreButton = getByText(messageShowMore).closest('button');

    expect(showMoreButton).not.toBeNull();

    if (showMoreButton) {
      expect(mockOnRelatedArticlesShowMoreClick).toHaveBeenCalledTimes(0);
      fireEvent.click(showMoreButton);
      expect(mockOnRelatedArticlesShowMoreClick).toHaveBeenCalledTimes(1);
    }
  });

  it('Should display the style 1 by default', async () => {
    const { queryByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(queryByText(messageTitle)).toBeNull();
  });

  it('Should display the style 2 when style="secondary"', async () => {
    const { getByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          style="secondary"
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() =>
            mockOnGetRelatedArticles(NUMBER_OF_ARTICLES)
          }
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(getByText(messageTitle)).not.toBeNull();
  });

  it('It should display the error text if the onGetRelatedArticles fails', async () => {
    const { queryByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() => mockOnGetRelatedArticles(0)}
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(queryByText(messageEndpointErrorTitle)).not.toBeNull();
    expect(queryByText(messageEndpointErrorDescription)).not.toBeNull();
  });

  it('If display the error screen and the user clicks the button "try again", It should execute onGetRelatedArticles again', async () => {
    const { queryByText, getByText } = render(
      <IntlProvider locale="en">
        <RelatedArticles
          routeGroup="test"
          routeName="test"
          onGetRelatedArticles={() => mockOnGetRelatedArticles(0)}
          onRelatedArticlesListItemClick={mockOnRelatedArticlesListItemClick}
          onRelatedArticlesShowMoreClick={mockOnRelatedArticlesShowMoreClick}
        />
      </IntlProvider>,
    );
    expect(mockOnGetRelatedArticles).toHaveBeenCalledTimes(1);

    jest.advanceTimersByTime(200);

    await waitFor(() => expect(mockOnGetRelatedArticles).toHaveBeenCalled());

    expect(queryByText(messageEndpointErrorTitle)).not.toBeNull();
    expect(queryByText(messageEndpointErrorDescription)).not.toBeNull();

    const tryAgainButton = getByText(messageEndpointErrorButtonLabel).closest(
      'button',
    );

    expect(tryAgainButton).not.toBeNull();
    if (tryAgainButton) {
      expect(mockOnGetRelatedArticles).toHaveBeenCalledTimes(1);
      fireEvent.click(tryAgainButton);
      expect(mockOnGetRelatedArticles).toHaveBeenCalledTimes(2);
    }
  });
});
