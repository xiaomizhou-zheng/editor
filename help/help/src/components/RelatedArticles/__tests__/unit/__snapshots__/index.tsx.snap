// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedArticles Should match snapshot 1`] = `
.emotion-23 {
  position: relative;
}

.emotion-3 {
  position: relative;
  box-sizing: border-box;
  padding: 8px;
  display: block;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  color: var(--ds-text-subtlest,#6B778C);
  border-radius: 3px;
  background-color: var(--ds-background-neutral-subtle,transparent);
  border: 0;
  padding: 8px;
  margin-bottom: 0;
}

.emotion-3:hover,
.emotion-3:focus,
.emotion-3:visited,
.emotion-3:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  outline-offset: none;
}

.emotion-3:focus {
  box-shadow: var(--ds-border-focused,#4C9AFF) 0px 0px 0px 2px inset;
}

.emotion-3:hover {
  background-color: var(--ds-background-neutral-subtle-hovered,#EBECF0);
}

.emotion-3:active {
  background-color: var(--ds-background-neutral-subtle-pressed,#DEEBFF);
}

.emotion-1 {
  width: 100%;
  white-space: nowrap;
}

.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--ds-text,#172B4D);
  font-size: 14px;
  font-weight: 600;
  display: inline-block;
  line-height: 20px;
  white-space: normal;
  overflow-x: hidden;
  margin-bottom: 8px;
}

.emotion-2 {
  display: block;
  line-height: 20px;
  color: var(--ds-text-subtle,#505F79);
  margin: 0;
}

.emotion-22 {
  padding: 8px 0;
}

.emotion-22 span {
  margin: 0;
}



.emotion-21,
.emotion-21:hover,
.emotion-21:active,
.emotion-21:focus,
.emotion-21:visited,
.emotion-21:disabled,
.emotion-21[disabled] {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  border-width: 0;
  border-radius: 3px;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: inherit;
  font-style: normal;
  font-family: inherit;
  font-weight: 500;
  max-width: 100%;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47, 0.03, 0.49, 1.38);
  transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47, 0.03, 0.49, 1.38);
  white-space: nowrap;
  background: none;
  color: var(--ds-link, #0052CC)!important;
  cursor: pointer;
  height: 1.7142857142857142em;
  line-height: 1.7142857142857142em;
  padding: 0 10px;
  vertical-align: middle;
  width: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline: none;
}

.emotion-21::-moz-focus-inner,
.emotion-21:hover::-moz-focus-inner,
.emotion-21:active::-moz-focus-inner,
.emotion-21:focus::-moz-focus-inner,
.emotion-21:visited::-moz-focus-inner,
.emotion-21:disabled::-moz-focus-inner,
.emotion-21[disabled]::-moz-focus-inner {
  border: 0;
  margin: 0;
  padding: 0;
}

.emotion-20 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  margin: 0 2px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
  class="emotion-23"
>
  <div>
    <a
      aria-disabled="false"
      class="emotion-3"
      href=""
      role="button"
    >
      <div
        class="emotion-1"
      >
        <span
          class="emotion-0"
        >
          mockArticle title
          
        </span>
      </div>
      <p
        class="emotion-2"
      >
        mockArticle description
      </p>
    </a>
    <a
      aria-disabled="false"
      class="emotion-3"
      href=""
      role="button"
    >
      <div
        class="emotion-1"
      >
        <span
          class="emotion-0"
        >
          mockArticle title
          
        </span>
      </div>
      <p
        class="emotion-2"
      >
        mockArticle description
      </p>
    </a>
    <a
      aria-disabled="false"
      class="emotion-3"
      href=""
      role="button"
    >
      <div
        class="emotion-1"
      >
        <span
          class="emotion-0"
        >
          mockArticle title
          
        </span>
      </div>
      <p
        class="emotion-2"
      >
        mockArticle description
      </p>
    </a>
  </div>
  <div
    aria-hidden="true"
    class="rah-static rah-static--height-zero"
    style="height: 0px; overflow: hidden;"
  >
    <div
      style="display: none;"
    >
      <a
        aria-disabled="false"
        class="emotion-3"
        href=""
        role="button"
      >
        <div
          class="emotion-1"
        >
          <span
            class="emotion-0"
          >
            mockArticle title
            
          </span>
        </div>
        <p
          class="emotion-2"
        >
          mockArticle description
        </p>
      </a>
      <a
        aria-disabled="false"
        class="emotion-3"
        href=""
        role="button"
      >
        <div
          class="emotion-1"
        >
          <span
            class="emotion-0"
          >
            mockArticle title
            
          </span>
        </div>
        <p
          class="emotion-2"
        >
          mockArticle description
        </p>
      </a>
    </div>
  </div>
  <div
    class="emotion-22"
  >
    <button
      class="emotion-21"
      tabindex="0"
      type="button"
    >
      <span
        class="emotion-20"
      >
        Show 2 more articles
      </span>
    </button>
  </div>
</div>
`;
